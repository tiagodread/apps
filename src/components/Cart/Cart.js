export default class Cart {
  items = [];

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc + item.quantity * item.product.price;
    }, 0);
  }

  add(item) {
    this.items.push(item);
  }
}
