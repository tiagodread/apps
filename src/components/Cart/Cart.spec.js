import Cart from "./Cart";

describe("Cart", () => {
  let cart;
  let productBase;

  beforeEach(() => {
    cart = new Cart();
    productBase = {
      product: {
        name: "Macbook Pro",
        price: 12569.0,
      },
      quantity: 2,
    };
  });

  afterEach(() => {
    cart = undefined;
  });

  describe("Add and remove items", () => {
    it("Should return 0 from getTotalCart when a new cart is created", () => {
      expect(cart.getTotalCart().getAmount()).toEqual(0);
    });

    it("Should multiply the product price and quantity to obtain the total", () => {
      cart.add(productBase);
      expect(cart.getTotalCart().getAmount()).toEqual(25138.0);
    });

    it("Should multiply the product price and quantity to obtain the total even for duplicated products", () => {
      cart.add(productBase);
      cart.add({
        product: {
          name: "Macbook Pro",
          price: 12569,
        },
        quantity: 1,
      });
      expect(cart.getTotalCart().getAmount()).toEqual(37707);
      expect(cart.getItemQuantity(productBase)).toEqual(3);
    });

    it("Should ensure one product exists at a time", () => {
      cart.add(productBase);
      expect(cart.getItemQuantity(productBase)).toEqual(2);
      cart.add(productBase);
      expect(cart.getItemQuantity(productBase)).toEqual(4);
      expect(cart.getTotalCart().getAmount()).toEqual(50276);
      expect(cart.getCart()).toEqual([productBase]);
    });

    it("Should add an item, remove it and the total be updated", () => {
      const product2 = {
        product: {
          name: "Iphone 13 Pro",
          price: 6599,
        },
        quantity: 1,
      };
      cart.add(productBase);
      cart.add(product2);
      expect(cart.getItemQuantity(productBase)).toEqual(2);
      expect(cart.getItemQuantity(product2)).toEqual(1);
      expect(cart.getCart()).toEqual([productBase, product2]);
      expect(cart.getTotalCart().getAmount()).toEqual(31737);
      cart.remove(product2);
      expect(cart.getItemQuantity(product2)).toEqual(0);
      expect(cart.getTotalCart().getAmount()).toEqual(25138);
    });
  });

  describe("Checkout", () => {
    it("Should return an object with items and total", () => {
      cart.add(productBase);
      cart.add({
        product: {
          name: "Iphone 13 Pro",
          price: 6599,
        },
        quantity: 2,
      });
      cart.add({
        product: {
          name: "Apple Watch Series 7",
          price: 5022.0,
        },
        quantity: 5,
      });
      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotalCart().getAmount()).toEqual(63446);
      expect(cart.checkout()).toMatchSnapshot();
      expect(cart.getTotalCart().getAmount()).toEqual(0);
    });
  });
});
