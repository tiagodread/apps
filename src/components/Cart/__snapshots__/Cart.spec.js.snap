// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Checkout Should return an object with items and total 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "name": "Macbook Pro",
        "price": 12569,
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "name": "Iphone 13 Pro",
        "price": 6599,
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "name": "Apple Watch Series 7",
        "price": 5022,
      },
      "quantity": 5,
    },
  ],
  "total": 63446,
}
`;

exports[`Cart Checkout Should return an object with items and total 2`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "name": "Macbook Pro",
        "price": 12569,
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "name": "Iphone 13 Pro",
        "price": 6599,
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "name": "Apple Watch Series 7",
        "price": 5022,
      },
      "quantity": 5,
    },
  ],
  "total": 63446,
}
`;
